File: JavaScript/Programmers/Level 1/두 개 뽑아서 더하기.js
// numbers에서 2개의 수를 뽑아 만들수 있는 모든 조합을 구함
// 각 조합마다 합을 구하고 정렬

// function combinations(numbers, pick) { // 조합 구현
//     let result = [];
    
//     if (pick === 1) {
//         return numbers.map(number => [number]);
//     }
    
//     numbers.forEach((fixed, idx) => {
//         const new_nums = numbers.slice(idx + 1);
//         const combination = combinations(new_nums, pick - 1);
//         combination.forEach(item => result.push([fixed, ...item]));
//     });
    
//     return result;
// }

// function solution(numbers) {
//     const answer = combinations(numbers, 2).map(item => item[0] + item[1]); // 각각의 조합마다 합을 구함
//     return [...new Set(answer)].sort((a, b) => a - b); // new Set()을 이용하여 중복을 제거하고 전개문법을 이용하여 배열로 만든 후 오름차순 정렬
// }

// 모든 조합을 구하고, 모든 조합 경우에 대해 합을 구해야 하기 때문에
// 2개의 수를 뽑아 더하는 것이므로 중첩 for문을 이용하여 한번에 구하는 것으로 수정

function solution(numbers) {
    const answer = [];
    for (let i = 0; i < numbers.length; i++) {
        for (let j = i + 1; j < numbers.length; j++) {
            answer.push(numbers[i] + numbers[j]); // 2개의 수를 뽑아 합을 push
        }
    }
    
    return [...new Set(answer)].sort((a, b) => a - b); // new Set()을 이용하여 중복을 제거하고 전개문법을 이용하여 배열로 만든 후 오름차순 정렬
}

File: Python/Programmers/Level 1/두 개 뽑아서 더하기.py
# from itertools import combinations

# def solution(numbers):
#     combi = list(combinations(numbers, 2)) # 2개 조합 만들기
#     answer = [a + b for a, b in combi] # 각 조합에 대해 모든 덧셈 구하기
#     answer = list(set(answer)) # 중복값 제거
#     return sorted(answer) # 오름차순 정렬

# from itertools import combinations

# def solution(numbers):
#     answer = list(map(sum, combinations(numbers, 2))) # map을 이용하여 2개 조합 덧셈 리스트 만들기
#     return sorted(list(set(answer))) # 중복제거 및 오름차순 정렬

# from itertools import combinations

# def solution(numbers):
#     return sorted(set(map(sum, combinations(numbers, 2)))) # map 결과는 set으로 만들고 오름차순 정렬

# 백트레킹 풀이, dfs 알고리즘 이용

def solution(numbers):
    def dfs(case, pick, idx): # dfs 정의
        if len(case) == pick: # 뽑은 경우가 pick와 같다면
            answer.add(sum(case)) # answer에 case의 합 추가
            return # 재귀 탈출
        
        for i in range(idx, len(numbers)): # numbers의 모든 요소를 순회하면서 모든 경우 탐색
            dfs(case + [numbers[i]], pick, i + 1) # dfs 탐색
    
    answer = set()
    dfs([], 2, 0)
    return sorted(answer)
